# Astronomy Database Query
(15 April 2026)



ADQ takes advantage of the Application Programming Interface (API) from the European Space Agency (ESA) to access Gaia telescope information and uses database queries to access specific information.  This project directly builds off the **PIMA School (24 March 2024)** independant study and utilizes the same environment build, with the addition of adding the **astroquery** library.  Instructions to add astroquery are both below, and in a separate document on how to set up your environment.

The ESA's Gaia Project originally started as the GAIA Project: **Global Astrometric Interferometer of Astrophysics**, reflecting the original interferomy optical technique they planned to use, but over the course of the project's development that changed, and GAIA was renamed Gaia.

Launched 19 December 2013, Gaia conducted a Lissajous orbit in the Sun-Earth L2 Lagrangian point, its mission to measure positions, distances and motions of stars, and to calculate positions of exoplanets.  Gaia's mission statement directly supports this project's investigation into parallax and stellar distances.  This is the exact same orbital plan the James Webb telescope utilizes, as it takes minimal fuel to maintain its position.  

Gaia ran out of fuel in 2025, no longer able to maintain its L2 orbital position, and was moved into a safe heliocentric (or circumsolar) orbit and deactivated 27 March 2025.  All the data had been completely downloaded January 2025, with processing expected to continue until 2030.

Gaia is not unique in our ability to access data via API.  A list of other accessible projects will be given in the References for future investigations.  Projects may have modified or developed their own API; you will need to reference that project directly to determine how to access it.

In addition to the elements we added to create the 'pima' environment in a prior AST296LB, we need to add astroquery.  
*I am going to modify this to include the necessary instructions as well as append information to the pass-out we got from the last class to set up a computer from absolute zero*

Open your terminal, activate the pima environment by entering in:

  **conda activate pima**

Once that resolves, enter the following:

  **python -m pip install -U --pre astroquery[all]**

This command will install astroquery, along with all its dependancies ([all] argument).


Or, you can also use conda to install:

  **conda install -c conda-forge astroquery**

### Import modules

In [1]:
# We need to initiate and import the appropriate modules in order to format and access all of our data.
%matplotlib inline

# Standard astronomy imports
import numpy as np
import matplotlib.pyplot as plt
from astropy.io import fits
from astropy.wcs import WCS
import sep

# Gaia-specific imports
from astroquery.gaia import Gaia

The archive is unstable and may perform below expectations. If launching multiple, consecutive, heavy queries through Python, please space them out (e.g., using sleep(1)) to avoid overloading the system. Please contact the Gaia helpdesk in case of questions (https://www.cosmos.esa.int/web/gaia/gaia-helpdesk). Workaround solutions for the issues following the December 2025 infrastructure upgrade: https://www.cosmos.esa.int/web/gaia/news#WorkaroundArchive

The GACS will be under maintenance on 15 April from 11:00 to 13:00 CEST approximately, for the upgrade to version 3.12.0. Apologies for the inconveniences.


### Log into GAIA
#### (optional)

You can register for an account at **https://www.cosmos.esa.int/web/gaia-users/register**.

It is not necessary to do this; what logging in does is allows you to save various tables to share with others, etc.  you can still access everything you need without this however.

We are going to use:

**Gaia.login()**

You will be prompted for your user name and password as we didn't pass any arguments.  You can set this format to include passing your information if you feel comfortable doing so, but your system's password manager should handle this for you as well.

**Gaia.login(user='userName', password='userPassword')**



In [5]:
#Log into Gaia with your COSMOS user account
Gaia.login()

INFO: Login to gaia TAP server [astroquery.gaia.core]


User:  dvancejr
Password:  ········


401 Error 401:
<!doctype html><html lang="en"><head><title>HTTP Status 401 – Unauthorized</title><style type="text/css">body {font-family:Tahoma,Arial,sans-serif;} h1, h2, h3, b {color:white;background-color:#525D76;} h1 {font-size:22px;} h2 {font-size:16px;} h3 {font-size:14px;} p {font-size:12px;} a {color:black;} .line {height:1px;background-color:#525D76;border:none;}</style></head><body><h1>HTTP Status 401 – Unauthorized</h1><hr class="line" /><p><b>Type</b> Status Report</p><p><b>Message</b> Bad Credentials</p><p><b>Description</b> The request has not been applied to the target resource because it lacks valid authentication credentials for that resource.</p><hr class="line" /><h3>Apache Tomcat/10.1.39</h3></body></html>


ERROR: Error logging in TAP server [astroquery.gaia.core]


# Accessing the Database and pulling table information

For this project I am focusing on Gaia Data Release 3.  More information on the different releases can be found at:
https://www.cosmos.esa.int/web/gaia/release#:~:text=On%20this%20webpage%2C%20the%20target%20release%20scenario%20is,will%20be%20released%20only%20with%20the%20final%20catalogue.

#### Load Tables

Load the names of the tables in the Gaia database using the command

**Gaia.load_tables(only_names=True)**

The option **only_names=True** returns just the information about the tables, not the actual contents.

    

In [2]:
tables = Gaia.load_tables(only_names=True)

INFO: Retrieving tables... [astroquery.utils.tap.core]
INFO: Parsing tables... [astroquery.utils.tap.core]
INFO: Done. [astroquery.utils.tap.core]


And now we can diplay the table names:

In [3]:
for table in tables:
    print(table.name)

external.apassdr9
external.catwise2020
external.gaiadr2_astrophysical_parameters
external.gaiadr2_geometric_distance
external.gaiaedr3_distance
external.gaiaedr3_gcns_main_1
external.gaiaedr3_gcns_rejected_1
external.gaiaedr3_spurious
external.gaia_eso_survey
external.galex_ais
external.lamost_dr9_lrs
external.lamost_dr9_mrs
external.ravedr5_com
external.ravedr5_dr5
external.ravedr5_gra
external.ravedr5_on
external.ravedr6
external.sdssdr13_photoprimary
external.skymapperdr1_master
external.skymapperdr2_master
external.tmass_xsc
external.xgboost_table1
external.xgboost_table2
gaiadr1.aux_qso_icrf2_match
gaiadr1.ext_phot_zero_point
gaiadr1.allwise_best_neighbour
gaiadr1.allwise_neighbourhood
gaiadr1.gsc23_best_neighbour
gaiadr1.gsc23_neighbourhood
gaiadr1.ppmxl_best_neighbour
gaiadr1.ppmxl_neighbourhood
gaiadr1.sdss_dr9_best_neighbour
gaiadr1.sdss_dr9_neighbourhood
gaiadr1.tmass_best_neighbour
gaiadr1.tmass_neighbourhood
gaiadr1.ucac4_best_neighbour
gaiadr1.ucac4_neighbourhood
gaiadr1.u

#### Accessing a table's meta data

This results in a lot of tables.  We can next query a specific table (**load_table**, not **load_tables**) and see its specific descriptive meta data:

In [4]:
meta = Gaia.load_table('gaiadr3.gaia_source')
meta

print(meta)

TAP Table name: gaiadr3.gaia_source
Description: This table has an entry for every Gaia observed source as published with this data release. It contains the basic source parameters, in their final state as processed by the Gaia Data Processing and Analysis Consortium from the raw data coming from the spacecraft. The table is complemented with others containing information specific to certain kinds of objects (e.g.~Solar--system objects, non--single stars, variables etc.) and value--added processing (e.g.~astrophysical parameters etc.). Further array data types (spectra, epoch measurements) are presented separately via Datalink resources.
Size (bytes): 3646930329600
Num. columns: 152


### Columns

Gaia3 has 152 columns.  We've already loaded the meta data, so another loop will give us the columns in the table.

In [5]:
for column in meta.columns:
    print(column.name)

solution_id
designation
source_id
random_index
ref_epoch
ra
ra_error
dec
dec_error
parallax
parallax_error
parallax_over_error
pm
pmra
pmra_error
pmdec
pmdec_error
ra_dec_corr
ra_parallax_corr
ra_pmra_corr
ra_pmdec_corr
dec_parallax_corr
dec_pmra_corr
dec_pmdec_corr
parallax_pmra_corr
parallax_pmdec_corr
pmra_pmdec_corr
astrometric_n_obs_al
astrometric_n_obs_ac
astrometric_n_good_obs_al
astrometric_n_bad_obs_al
astrometric_gof_al
astrometric_chi2_al
astrometric_excess_noise
astrometric_excess_noise_sig
astrometric_params_solved
astrometric_primary_flag
nu_eff_used_in_astrometry
pseudocolour
pseudocolour_error
ra_pseudocolour_corr
dec_pseudocolour_corr
parallax_pseudocolour_corr
pmra_pseudocolour_corr
pmdec_pseudocolour_corr
astrometric_matched_transits
visibility_periods_used
astrometric_sigma5d_max
matched_transits
new_matched_transits
matched_transits_removed
ipd_gof_harmonic_amplitude
ipd_gof_harmonic_phase
ipd_frac_multi_peak
ipd_frac_odd_win
ruwe
scan_direction_strength_k1
scan_di

#### Specific Data Query

At this point we still have too much information to do anything meaningful.  What we can now do is ask the database (query) to give us specific information.  The capitalized words are specific query commands:
* SELECT: We are selecting data, not adding or modifying it
* TOP:    This limits us to only the first X rows of the table.  For this initial forray I only want to pull 10 values.
          We run into processing time concerns when querying for large numbers of entries.  When pulling 2000 or less entries, a synchroneous
          search (one right after the other) is acceptable, but anything more than that will result in significant time issues.  If we were to
          query for a larger sample, an asynchroneous search would be the way to go, meaning the process wouldn't wait for the initial database query
          to resolve completely before moving on to the next task, allowing us to start processing data while the entire search continues in the
          background.
* FROM:   This defines which of the **tables** we listed above that we want to look at.

My particular interest is parallax.  Looking at the column names I can see some interesting information I would like to get:

In [6]:
query1 = """SELECT
            TOP 20
            source_id, ra, dec, parallax, parallax_error, parallax_over_error
            FROM gaiadr3.gaia_source
         """

In order to run this query we use **launch_job**:

In [7]:
job = Gaia.launch_job(query1)

We now have an object named job that we can get the metadata as well as printing out the table.

In [8]:
print(job)

<Table length=20>
        name         dtype  unit                            description                             n_bad
------------------- ------- ---- ------------------------------------------------------------------ -----
          source_id   int64      Unique source identifier (unique within a particular Data Release)     0
                 ra float64  deg                                                    Right ascension     0
                dec float64  deg                                                        Declination     0
           parallax float64  mas                                                           Parallax     2
     parallax_error float32  mas                                         Standard error of parallax     2
parallax_over_error float32                                  Parallax divided by its standard error     2
Jobid: None
Phase: COMPLETED
Owner: None
Output file: 2770f7f3-21b7-11f1-b85c-bc97e148b76b-O-result.vot.gz
Results: None


In [9]:
results = job.get_results()
results

source_id,ra,dec,parallax,parallax_error,parallax_over_error
,deg,deg,mas,mas,
int64,float64,float64,float64,float32,float32
549755818112,45.04828232129832,0.04825396034378256,1.5834770072004039,0.03442545,45.99728
2267742736896,45.079030047492104,0.08648916582061868,-0.10031703855761,0.6790846,-0.14772393
3508989119232,45.05595381188336,0.14611201290768125,1.169073322327338,0.029833404,39.18672
4741643961344,44.87719515522742,0.11741445611013862,1.003909573828523,1.076071,0.9329399
4810363437952,44.91029694323249,0.11475560366775076,1.7671974454860495,0.2355501,7.502427
6047314116736,44.915239585539965,0.158358051259343,0.3244697504024929,0.16274305,1.9937549
6944962925824,44.975821556411596,0.1878141021101549,0.9904046280949322,0.019033762,52.0341
8418136080384,45.03740766252847,0.26161351973494235,1.8108672639651335,0.13945648,12.985178


*At this point I want to start isolating the parallax data we have pulled to start calculating parallax.  I am also wanting to be able to use source_id to identify the specific object name we are looking at.  That is for next week!!*

#### Pulling information and doing calculations

Our results table is of the type:

In [10]:
type(results)

astropy.table.table.Table

What we are going to do is reduce the overall table into columns, which we can then access like we would a list.

In [11]:
source = results['source_id']
parallax = results['parallax']

Now we throw source and parallax into a loop.  Some values were given as '--', and some parallax were given as a negative.  Neither are of value to us, so the loop will only process values, and print a message if the values given are a negative.  Negative values happen when confusion occurs in the observation of a source, such as sources being too close together, too faint, or in a dense region.  Xavier Luri et all discuss the value of negative or large-error parallax in Gaia Data Release 2: Using Gaia Parallaxes, 25 April 2018: https://arxiv.org/abs/1804.09376

I have not had a chance to process this paper, and it opens up a whole new avenue of study that may or may not be beyond the scope of this project.  

For now, I am choosing to eliminate negative values and simply address each source individually.

In [12]:
# Set accumulator to 0
index = 0

# Loop through the 10 values we have
while index < 20:

    # Only calculate parallax that has values; skip if '--' is the value
    if parallax[index] != '--':
        pDistance = 1 / (parallax[index] * .001) # The parallax formula is in arc-seconds, but the information from Gaia is in milliarc-seconds
        lyDistance = pDistance * 3.26
            
        if pDistance <= 0: # If parallax is a negative print an error message
            print(f'The parallax data for {source[index]} is uncertain; refer to Luri et all, 2018.')

        else: # Display the source's distance in parsec and lightyear.
            print(f'The distance to {source[index]} is {pDistance:.3f} parsecs or {lyDistance:.3f} lightyears')
    index += 1

The distance to 549755818112 is 631.522 parsecs or 2058.761 lightyears
The parallax data for 2267742736896 is uncertain; refer to Luri et all, 2018.
The distance to 3508989119232 is 855.378 parsecs or 2788.533 lightyears
The distance to 4741643961344 is 996.106 parsecs or 3247.304 lightyears
The distance to 4810363437952 is 565.868 parsecs or 1844.729 lightyears
The distance to 6047314116736 is 3081.951 parsecs or 10047.162 lightyears
The distance to 6944962925824 is 1009.688 parsecs or 3291.584 lightyears
The distance to 8418136080384 is 552.222 parsecs or 1800.242 lightyears
The distance to 8834748568832 is 427.651 parsecs or 1394.143 lightyears
The parallax data for 12678743468032 is uncertain; refer to Luri et all, 2018.
The distance to 13125420068480 is 647.722 parsecs or 2111.572 lightyears
The distance to 15633681152896 is 529.878 parsecs or 1727.404 lightyears
The distance to 16973710962944 is 1658.989 parsecs or 5408.303 lightyears
The distance to 17527761792640 is 622.807 par